packages/wonder-stuff-i18n/src/plugins/i18n-plugin.js

```diff
@@ -0,0 +1,410 @@
+// @flow
+/**
+ * A plugin to create translated versions of Webpack builds. After Webpack compiles
+ * all the JS bundles, this plugin goes through all the generated assets and
+ * does the following for each locale we want to build and for each file in
+ * the emitted assets:
+ *
+ * - Downloads the translated string JSON files from GCS.
+ * - Goes through each asset and transates its contents into each locale that
+ *   we support.
+ * - Updates any runtime files to ensure that the correct hashes are used in
+ *   the filenames.
+ * - Updates the manifest files for those locales to point to the right files.
+ *
+ * To use this plugin, pass it into the `plugins` array of your Webpack config:
+ *
+ *      plugins: [...others, new I18nPlugin({locales: ["en", "es"]})]
+ *
+ */
+import path from "path";
+import util from "util";
+import {eachLimit} from "async";
+
+import * as I18nUtils from "../utils/i18n-utils.js";
+import {
+    hashFileContents,
+    extractHashFromFileName,
+    localizeFile,
+} from "../utils/localize-file.js";
+import {getEmojiForLocale} from "../utils/emoji-for-locale.js";
+
+import type {TranslatedLocaleStrings} from "../utils/i18n-utils.js";
+
+// NOTE(jeresig): If the NPM async module was using v3 we could just use it
+// directly, instead of having to use promisify.
+const asyncEachLimit = util.promisify(eachLimit);
+
+/**
+ * Add a new asset to the Webpack compilation assets.
+ *
+ * @param {Object} assets the compilation assets object from Webpack
+ * @param {string} relFilePath a file path relative to the Webpack output
+ *   directory.
+ */
+const addAsset = (assets, relFilePath, data) => {
+    assets[relFilePath] = {
+        source() {
+            return data;
+        },
+        size() {
+            return data.length;
+        },
+    };
+};
+
+export type TranslatedStrings = {
+    [locale: string]: TranslatedLocaleStrings,
+};
+
+type IncomingOptions = {
+    locales: Array<string>,
+    silent?: boolean,
+    timing?: {
+        start: (string) => void,
+        end: (string) => void,
+        ...
+    },
+    getI18nStrings: (locale: string) => Promise<TranslatedLocaleStrings>,
+    getLocalePath?: (locale: string) => string,
+    shouldLocalizeAsset?: (assetName: string) => boolean,
+    shouldProcessAsset?: (assetName: string) => boolean,
+    shouldLocalizeManifest?: (assetName: string, locale: string) => boolean,
+};
+
+type Options = {
+    locales: Array<string>,
+    silent?: boolean,
+    timing?: {
+        start: (string) => void,
+        end: (string) => void,
+        ...
+    },
+    getI18nStrings: (locale: string) => Promise<TranslatedLocaleStrings>,
+    getLocalePath: (locale: string) => string,
+    shouldLocalizeAsset: (assetName: string) => boolean,
+    shouldProcessAsset: (assetName: string) => boolean,
+    shouldLocalizeManifest: (assetName: string, locale: string) => boolean,
+};
```

Comment:
I think it would be easier to see the differences (and maintain the consistencies) between these if `IncomingOptions` was defined in terms of `Options`, e.g.
```
type IncomingOptions = {
   ...Options,
   getLocalePath?: Options['getLocalePath'],
   shouldLocalizeAsset?: Options['shouldLocalizeAsset'],
   shouldProcessAsset?: Options['shouldProcessAsset'],
   shouldLocalizeManifest?: Options['shouldLocalizeManifest'],
};
```

Reply:
Hmm... so I understand what you're suggesting, but I really want to have the docs be on the true "Options" object and have the "InternalOptions" be the one that's different (and undocumented). I'll upload what I have, but I don't think using `Options['getLocalePath']` will work in that case as it'll copy over the optional-ness of the property as well.